[Xtensa] Implement base CallConvention.

andreisfr opened the pull request:
Implement base Calling Convention functionality. Implement stack load/store register operations.

andreisfr committed:
[Xtensa] Implement base CallConvention.

Implement base Calling Convention functionality. Implement stack
load/store register operations.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -41,10 +41,273 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+
+      // The stack pointer offset is relative to the caller stack frame.
+      int FI = MFI.CreateFixedObject(ValVT.getStoreSize(), VA.getLocMemOffset(),
+                                     true);
```
```suggestion
                                     /*IsImmutable=*/true);
```

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -41,10 +41,273 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+      int FI = MFI.CreateFixedObject(ValVT.getStoreSize(), VA.getLocMemOffset(),
+                                     true);
+
+      if (Ins[VA.getValNo()].Flags.isByVal()) {
```
`Ins` cannot be indexed by `ValNo` because `ValNo` is the argument index, while `Ins` contains lowered arguments, potentially split into multiple pieces.
```suggestion
      if (Ins[i].Flags.isByVal()) {
```

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -23,5 +24,73 @@
+                                  MachineBasicBlock::iterator MBBI,
+                                  const DebugLoc &DL, MCRegister DestReg,
+                                  MCRegister SrcReg, bool KillSrc) const {
+  // when we are copying a phys reg we want the bits for fp
```
What is "the bits for fp"? Could you clarify the comment?

andreisfr replied:
Corrected commnet

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -23,5 +24,73 @@
+  DebugLoc DL = MBBI != MBB.end() ? MBBI->getDebugLoc() : DebugLoc();
+  unsigned LoadOpcode, StoreOpcode;
+  getLoadStoreOpcodes(RC, LoadOpcode, StoreOpcode, FrameIdx);
+  addFrameReference(BuildMI(MBB, MBBI, DL, get(StoreOpcode))
```
It would be easier to understand with BuildMI on the line before addFrameReference.

andreisfr replied:
Corrected

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -23,5 +24,73 @@
+  if (RC == &Xtensa::ARRegClass) {
+    LoadOpcode = Xtensa::L32I;
+    StoreOpcode = Xtensa::S32I;
+  } else
```
Missing braces.

andreisfr replied:
Corrected

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaOperators.td:
```diff
@@ -0,0 +1,13 @@
+//===----------------------------------------------------------------------===//
+// Node definitions
+//===----------------------------------------------------------------------===//
+def Xtensa_retflag: SDNode<"XtensaISD::RET", SDTNone,
```
```suggestion
def Xtensa_ret : SDNode<"XtensaISD::RET", SDTNone,
```

andreisfr replied:
Corrected

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp:
```diff
@@ -57,10 +58,47 @@ BitVector XtensaRegisterInfo::getReservedRegs(const MachineFunction &MF) const {
   return Reserved;
 }
 
+bool XtensaRegisterInfo::eliminateFI(MachineBasicBlock::iterator II,
```
Why don't just inline it?

andreisfr replied:
I merged functions and removed eliminateFI.

s-barannikov commented:
I wouldn't mind if this PR included LowerCall implementation, it won't be much different from LowerFormalArguments/LowerReturn. The tests could then just call a function with the same signature so that we can see how arguments are recieved and passed.
The PR should also include XtensaFrameLowering changes, at least a part of them, so that the tests are representative.
There should be tests for passing arguments on the stack.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.td:
```diff
@@ -238,6 +239,11 @@ def L32R : RI16_Inst<0x01, (outs AR:$t), (ins L32Rtarget:$label),
   let imm16 = label;
 }
 
+//extending loads
```
(not related to this PR)
I guess I'm late but still. It might make sense to split instruction definitions and patterns. E.g. extract pattern for i32 load from L32I definition and put it here together with other patterns for loads.
This will make instruction definitions more readable and patterns grouped so that you don't have to look through the entire file when searching for them.

andreisfr replied:
@s-barannikov , I prepared second commit with call lowering. Could you please take a look? We assume in the call lowering implementation that in most cases we should generate indirect call instruction and thus we need to load address constant from ConstantPool. So, I added base implementation of the ConstantPool to second commit. If it is not very complex to review both commits, I can merge them or drop second commit in other case.

s-barannikov replied:
It is very difficult indeed. I left some comments and will revisit once they are addressed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -41,10 +41,273 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+    RetValue = convertValVTToLocVT(DAG, DL, VA, RetValue);
+
+    // Chain and glue the copies together.
+    unsigned Reg = VA.getLocReg();
```
Register

andreisfr replied:
@arsenm , thank you very much for comments. Corrected

arsenm commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -23,5 +24,73 @@
+    BuildMI(MBB, MBBI, DL, get(Xtensa::OR), DestReg)
+        .addReg(SrcReg, getKillRegState(KillSrc))
+        .addReg(SrcReg, getKillRegState(KillSrc));
+  else
+    report_fatal_error("Impossible reg-to-reg copy");
```
reflow this into an assert (may not even be necessary to assert, I would assume the verifier catches the invalid regclass later)

gerekon commented:
Track in https://github.com/espressif/llvm-project/issues/4

andreisfr committed:
[Xtensa] Implement call lowering.

Implement lowering call operations. Implement stack pseudo
instructions ADJCALLSTACKDOWN, ADJCALLSTACKUP. Implement basic
support of the ConstantPool, integer and symbols constants placement
in literal section.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaUtils.cpp:
```diff
@@ -0,0 +1,60 @@
+  return isValidAddrOffset(Scale, Offset);
+}
+
+} // namespace llvm
```
Fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaUtils.h:
```diff
@@ -0,0 +1,27 @@
+// Check address offset for load/store instructions
+bool isValidAddrOffset(MachineInstr &MI, int64_t Offset);
+} // namespace llvm
+#endif // LLVM_LIB_TARGET_XTENSA_XTENSAUTILS_H
```
Fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaUtils.h:
```diff
@@ -0,0 +1,27 @@
+
+namespace llvm {
+// Check address offset for load/store instructions
+// The offset should be multiple of scale
```
Fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaCallingConv.td:
```diff
@@ -0,0 +1,28 @@
+// Xtensa return value calling convention
+//===----------------------------------------------------------------------===//
+def RetCC_Xtensa : CallingConv<[
+  CCIfType<[i1, i8, i16], CCPromoteToType<i32>>,
```
@s-barannikov, thank you very much for comments, could you please describe more precisely what should be done? 
I removed redundant line with f32 assignment to registers, but as I'm understand many other backends (like ARM for example) use similar approach with i1/i8/i16 types promotion.

s-barannikov replied:
In ARM, it is redundant too (it is only used in FastISel path).
SelectionDAGBuilder correctly handles promotions itself, inserting Assert[SZ]Ext / Truncate as necessary.

s-barannikov commented:
@andreisfr Please don't amend / force push future commits. This makes it difficult to track comments.

arsenm commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -148,7 +161,13 @@ struct XtensaOperand : public MCParsedAsmOperand {
-  bool isImm12m() const { return isImm(-2048, 2047); }
+  // Convert MOVI to literal load, when immediate is not in range (-2048, 2047)
+  bool isImm12m() const {
+    if (Kind == Immediate)
+      return true;
+
+    return false;
```
```suggestion
    return Kind == Immediate;
```

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -348,6 +367,67 @@ static SMLoc RefineErrorLoc(const SMLoc Loc, const OperandVector &Operands,
+  switch (Opcode) {
+  case Xtensa::L32R: {
+    const MCSymbolRefExpr *OpExpr =
+        (const MCSymbolRefExpr *)Inst.getOperand(1).getExpr();
```
Prefer static_cast

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -348,6 +367,67 @@ static SMLoc RefineErrorLoc(const SMLoc Loc, const OperandVector &Operands,
+    if (!Inst.getOperand(1).isExpr()) {
+      uint64_t ImmOp64 = Inst.getOperand(1).getImm();
+      int32_t Imm = ImmOp64;
+      if ((Imm < -2048) || (Imm > 2047)) {
```
!isInt<11>(Imm)?

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -348,6 +367,67 @@ static SMLoc RefineErrorLoc(const SMLoc Loc, const OperandVector &Operands,
+      Inst = TmpInst;
+      TS.emitLiteral(Sym, Value, IDLoc);
+    }
+  } break;
```
Weird break placement relative to brace

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -686,6 +767,58 @@ bool XtensaAsmParser::ParseInstruction(ParseInstructionInfo &Info,
+  const MCSymbolRefExpr *SE = dyn_cast<MCSymbolRefExpr>(Value);
+  if (!SE)
+    return Error(LiteralLoc, "literal label must be a symbol");
+  else {
```
No else after return

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -686,6 +767,58 @@ bool XtensaAsmParser::ParseInstruction(ParseInstructionInfo &Info,
 
+bool XtensaAsmParser::parseLiteralDirective(SMLoc L) {
+  MCAsmParser &Parser = getParser();
+  MCSymbol *Sym;
```
Sink this down to define on first use

andreisfr replied:
Corrected

arsenm commented on llvm/lib/Target/Xtensa/MCTargetDesc/XtensaTargetStreamer.cpp:
```diff
@@ -0,0 +1,87 @@
+  std::string SectionName;
+  if (Pos != std::string::npos) {
+    if (Pos > 0)
+      SectionName = CSectionName.substr(0, Pos + 5);
```
can you use consume_front(".text") instead of maintaining this + 5

arsenm commented on llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:
```diff
@@ -25,12 +29,193 @@
+
+  XtensaConstantPoolSymbol *XtensaSym = cast<XtensaConstantPoolSymbol>(ACPV);
+  const char *Sym = XtensaSym->getSymbol();
+  std::string SymName(Sym);
```
SmallString + svector_ostream?

andreisfr replied:
I used SmallString+svector_ostream to print literal values

arsenm commented on llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:
```diff
@@ -25,12 +29,193 @@
+  if (CP.empty())
+    return;
+
+  for (unsigned i = 0, e = CP.size(); i != e; ++i) {
+    const MachineConstantPoolEntry &CPE = CP[i];
```
range loop

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:
```diff
@@ -25,12 +29,193 @@
+        MCSectionELF *CS =
+            (MCSectionELF *)getObjFileLowering().SectionForGlobal(&F, TM);
+        std::string CSectionName = CS->getName().str();
+        std::size_t Pos = CSectionName.find(".text");
```
consume_front?

andreisfr replied:
fixed

andreisfr committed:
[Xtensa] Fix callconv, constantpool, etc

Fix type promotion in callconv implementation. Minor fixes in
constantpool, targetstreamer and asmparser. Fix minor format issues.

andreisfr committed:
[Xtensa] Fix asm printer.

Use SmallString instead of std::string.

arsenm commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -348,6 +362,68 @@ static SMLoc RefineErrorLoc(const SMLoc Loc, const OperandVector &Operands,
+    XtensaMCExpr::VariantKind Kind = XtensaMCExpr::VK_Xtensa_None;
+    const MCExpr *NewOpExpr = XtensaMCExpr::create(OpExpr, Kind, getContext());
+    Inst.getOperand(1).setExpr(NewOpExpr);
+  } break;
```
Formatting is off, break before brace

andreisfr replied:
fixed

arsenm commented on llvm/test/MC/Xtensa/Core/invalid.s:
```diff
@@ -4,10 +4,6 @@ LBL0:
 # Out of range immediates
 
-# imm12m
-movi a1, 3000
```
Lost the range verification test?

andreisfr replied:
I removed this test, because constant pool is implemented now. The Xtensa asm allows mnemonics like "movi a1, imm" even if "imm" is out of range, we can just replace it with loading contant from constant pool

arsenm commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -686,6 +763,57 @@ bool XtensaAsmParser::ParseInstruction(ParseInstructionInfo &Info,
+
+  const MCSymbolRefExpr *SE = dyn_cast<MCSymbolRefExpr>(Value);
+
+  if (!SE)
+    return Error(LiteralLoc, "literal label must be a symbol");
+
+  if (Parser.parseToken(AsmToken::Comma, "expected comma"))
+    return true;
+
+  SMLoc OpcodeLoc = getLexer().getLoc();
+  if (parseOptionalToken(AsmToken::EndOfStatement))
+    return Error(OpcodeLoc, "expected value");
```
I don't believe these error conditions are tested

andreisfr replied:
I added test for this checks

arsenm commented on llvm/lib/Target/Xtensa/XtensaConstantPoolValue.h:
```diff
@@ -0,0 +1,264 @@
+  }
+
+  void print(raw_ostream &O) const override;
+  void print(raw_ostream *O) const {
+    if (O)
+      print(*O);
+  }
```
This shouldn't be necessary?

andreisfr replied:
I removed it

arsenm commented on llvm/lib/Target/Xtensa/XtensaConstantPoolValue.h:
```diff
@@ -0,0 +1,264 @@
+    if (O)
+      print(*O);
+  }
+  void dump() const;
```
Should be wrapped in LLVM_ENABLE_DUMP guard like other dump functions

arsenm reviewed (commented):
lgtm with a few nits

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaConstantPoolValue.cpp:
```diff
@@ -0,0 +1,205 @@
+
+XtensaConstantPoolConstant::XtensaConstantPoolConstant(
+    const Constant *C, unsigned ID, XtensaCP::XtensaCPKind Kind)
+    : XtensaConstantPoolValue((Type *)C->getType(), ID, Kind), CVal(C) {}
```
This this cast necessary?

andreisfr replied:
I removed it

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaConstantPoolValue.cpp:
```diff
@@ -0,0 +1,205 @@
+using namespace llvm;
+
+XtensaConstantPoolValue::XtensaConstantPoolValue(
+    Type *Ty, unsigned id, XtensaCP::XtensaCPKind kind,
```
Capitalize variable names. In other methods, too.

andreisfr replied:
fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaConstantPoolValue.h:
```diff
@@ -0,0 +1,264 @@
+      if (Constants[i].isMachineConstantPoolEntry() &&
+          (Constants[i].getAlign() >= Alignment)) {
+        XtensaConstantPoolValue *CPV =
+            (XtensaConstantPoolValue *)Constants[i].Val.MachineCPVal;
```
Use `auto *` and `static_cast`.

andreisfr replied:
fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaConstantPoolValue.h:
```diff
@@ -0,0 +1,264 @@
+
+public:
+  static XtensaConstantPoolMBB *
+  Create(LLVMContext &C, const MachineBasicBlock *mbb, unsigned ID);
```
Capitalize parameter names.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaConstantPoolValue.h:
```diff
@@ -0,0 +1,264 @@
+/// XtensaConstantPoolJumpTable - Xtensa-specific constantpool values for Jump
+/// Table symbols.
+class XtensaConstantPoolJumpTable : public XtensaConstantPoolValue {
+  unsigned IDX; // Jump Table Index.
```
```suggestion
  unsigned Idx; // Jump Table Index.
```

andreisfr replied:
fixed

s-barannikov reviewed (commented):
I understand nothing in constant pools, so I've left only nitpicks.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -41,18 +52,507 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+  bool needs64BitAlign = (ValVT == MVT::i32 && OrigAlign == Align(8));
+  bool needs128BitAlign = (ValVT == MVT::i32 && OrigAlign == Align(16));
+
+  if (ValVT == MVT::i32 || ValVT == MVT::f32) {
```
```suggestion
  if (ValVT == MVT::i32) {
```

andreisfr replied:
fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -41,18 +52,507 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+  // Get a count of how many bytes are to be pushed on the stack.
+  unsigned NumBytes = CCInfo.getStackSize();
+
+  unsigned StackAlignment = TFL->getStackAlignment();
```
The method is deprecated, consider using `getStackAlign`.

andreisfr replied:
fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -41,18 +52,507 @@ XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
+    // Check if use node maybe lowered to the ADDMI instruction
+    SDNode &OpNode = *Op.getNode();
+    if ((OpNode.hasOneUse() && OpNode.use_begin()->getOpcode() == ISD::ADD) &&
+        (value >= -32768) && (value <= 32512) && ((value & 0xff) == 0))
```
Consider using isShiftedInt from MathExtras.h here and in other places.

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -23,5 +24,139 @@
 
 using namespace llvm;
 
+static inline const MachineInstrBuilder &
```
```suggestion
static const MachineInstrBuilder &
```

andreisfr replied:
fixed

s-barannikov commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -23,5 +24,139 @@
+
+  if (isInt<8>(Amount)) // addi sp, sp, amount
+    BuildMI(MBB, I, DL, get(Xtensa::ADDI), Reg).addReg(SP).addImm(Amount);
+  else { // Expand immediate that doesn't fit in 12-bit.
```
8 bit?
Missing braces around if.

andreisfr replied:
fixed

andreisfr committed:
[Xtensa] Improve TargetStreamer, code formatting.

Move some functionality from AsmPrinter to the TargetStreamer. Add test for
"literal" asm directive. Code formatting.

andreisfr commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -35,6 +38,12 @@ class XtensaAsmParser : public MCTargetAsmParser {
+    return static_cast<XtensaTargetStreamer &>(TS);
+  }
+
+  bool ParseDirective(AsmToken DirectiveID) override;
```
fixed

andreisfr commented on llvm/lib/Target/Xtensa/AsmParser/XtensaAsmParser.cpp:
```diff
@@ -686,6 +763,57 @@ bool XtensaAsmParser::ParseInstruction(ParseInstructionInfo &Info,
+    XtensaTargetStreamer &TS = this->getTargetStreamer();
+    TS.emitLiteralPosition();
+    Lex();
+    return false;
```
fixed

andreisfr commented on llvm/lib/Target/Xtensa/MCTargetDesc/XtensaTargetStreamer.cpp:
```diff
@@ -0,0 +1,89 @@
+                                          SMLoc L) {
+  MCContext &Context = getStreamer().getContext();
+  MCStreamer &OutStreamer = getStreamer();
+  MCSectionELF *CS = (MCSectionELF *)OutStreamer.getCurrentSectionOnly();
```
fixed

andreisfr commented on llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:
```diff
@@ -25,12 +29,193 @@
+    StringRef Modifier = ACPV->getModifierText();
+    LiteralStr << Modifier;
+
+    OutStreamer->emitRawText(StringRef(LiteralStr.str()));
```
I refactored XtensaTargetStreamer to implement some if/else code from AsmPrinter in XtensaAsmTargetStreamer and XtensaElfTargetStreamer, PTAL.

andreisfr committed:
[Xtensa] Minor fixes.

andreisfr commented:
@s-barannikov , may I ask you to review new changes?
